### PR TITLE
Producer consumer ocr

### PR DESCRIPTION
**Pull request by kingst**
Introduces the consumer / producer OCR. Highlights include:
- Significantly faster frame rate
- Name extraction
- Better expiry

### CODE REVIEW

**Comment by jaimejiyepark** [comment]
I tested it out with my phone and wow the legal name extraction is really cool

File: CardScan/Classes/CreditCardOcr/CreditCardOcrImplementation.swift [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    
+    init(dispatchQueue: DispatchQueue) {
+        self.dispatchQueue = dispatchQueue
+    }
+    
+    func recognizeCard(in fullImage: CGImage, roiRectangle: CGRect) -> CreditCardOcrPrediction {
+        preconditionFailure("This method must be overridden")
+    }
+    
+    func croppedImage(fullCardImage: CGImage, roiRectangle: CGRect) -> CGImage? {
```
Comment (jaimejiyepark) [inline CardScan/Classes/CreditCardOcr/CreditCardOcrImplementation.swift:37]: would it make sense to put these cropping functions in imageUtils?

File: CardScan/Classes/CreditCardOcr/AppleCreditCardOcr.swift [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        semaphore.wait()
+        let duration = -startTime.timeIntervalSinceNow
+        self.computationTime += duration
+        self.frames += 1
+
+        return CreditCardOcrPrediction(image: image, number: pan, expiryMonth: expiryMonth, expiryYear: expiryYear, name: name, computationTime: duration, numberBoxes: numberBox.map { [$0] }, expiryBoxes: expiryBox.map { [$0] }, nameBoxes: nameBox.map { [$0] })
+    }
+    
+    static func likelyName(_ text: String) -> String? {
+        let lettersAndSpace = text.reduce(true) { acc, value in
```
Comment (jaimejiyepark) [inline CardScan/Classes/CreditCardOcr/AppleCreditCardOcr.swift:61]: what's acc?

File: CardScan/Classes/CreditCardOcr/CreditCardOcrPrediction.swift [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            return nil
+        }
+
+        let result = regex.matches(in: string, range: NSRange(string.startIndex..., in: string))
+        
+        if result.count == 0 {
+            return nil
+        }
+        
+        let range1 = Range(result.first!.range(at: 1), in: string)!
```
Comment (jaimejiyepark) [inline CardScan/Classes/CreditCardOcr/CreditCardOcrPrediction.swift:60]: I know `result.first` is taken care of by the previous `result.count` conditional statement but for the range would it be safer to use a default value? Even though we know result will have something but the `!` gives me nightmares.

I'm not sure though.

File: CardScan/Classes/CreditCardOcr/CreditCardOcrImplementation.swift (hunk 1 above)
Comment (kingst) [reply CardScan/Classes/CreditCardOcr/CreditCardOcrImplementation.swift:37]: yeah it does, but let's do that later. There is still a ton of refactoring that we need to do.

File: CardScan/Classes/CreditCardOcr/AppleCreditCardOcr.swift (hunk 2 above)
Comment (kingst) [reply CardScan/Classes/CreditCardOcr/AppleCreditCardOcr.swift:61]: accumulator

File: CardScan/Classes/CreditCardOcr/CreditCardOcrPrediction.swift (hunk 3 above)
Comment (kingst) [reply CardScan/Classes/CreditCardOcr/CreditCardOcrPrediction.swift:60]: yikes! I'll fix that, nice catch.

**Pull request merged**